### PR TITLE
First pass of twizzler-object design.

### PR DESCRIPTION
**Pull request by dbittman**

### CODE REVIEW

**Comment by abenetopoulos** [comment]
👀

**Comment by PandaZ3D** [comment]
This looks great so far. A few comments/questions on the FOT entry, persistence, pointers, and `TxCell`.

- Why is the FOT entry reference counted, it's not clear at the moment what it could be used for.
- Throughout the document you avoid mentioning persistence. What is Twizzler's/this crate's take on this? I imagine a world where all computation is on in-memory objects, but some of those might want to be stored somewhere else (e.g. flash). Is it not the place for this crate to make this notion explicit? 
- Basically what are the invariants of Invariant Pointers? Are the non-invariant pointers somewhat like non-crash-safe (volatile) references? Does the system/runtime allow or restrict the construction of invariant pointers to volatile objects? 
- For `TxCell`, is the rationale there that requiring programmers to be explicit means not having to modify the compiler to make `transaction` log only `Foo.a`, relying on the type-system to do the work? As an alternative, could the macro system be used to parse and annotate the program so that only `Foo.a` is logged? Not sure if it's powerful enough to do something like that.

**Review by palvaro — COMMENTED** [review]
"A shared object is one that may be operated on by multiple *instances of a host*[^2]. " (and its footnote)

Although everything you say here makes sense, would it not be more parsimonious to say "A volatile object shares fate with a running kernel process; by contrast, a shared object may be operated on by multiple kernel processes, whether distributed in time (eg across reboots) or space (eg across hosts or devices)."

**Review by palvaro — COMMENTED** [review]
"The data structure ensures that mutation is possible only via transactions. [...]
 by implementing `!ObjSafe` for things like UnsafeCell."

blacklisting? how many such things are there?

**Comment by dbittman** [comment]
Updated based on our discussion on Friday. This changes some core framing towards a tighter coupling with twizzler-nando, with the following conceit:

Any mutable data should only be accessed (safely) via the twizzler-nando crate. Any object memory that is "always immutable" can be accessed via the twizzler-object crate.

The result is that the twizzler-object crate is smaller, but more well-defined.

File: text/0101-object.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+Twizzler object. This type exposes the following interfaces:
+
+```{rust}
+fn id(&self) -> ObjID;
+fn init_id(id: ObjID, prot: Protections, flags: InitFlags) -> Result<Self, InitError>;
+fn base(&self) -> &T;
+```
+
+Note that the base function returns an immutable reference to the base, and there is no safe way to
+get a mutable reference. This is because all mutation should be done via the twizzler-nando crate.
```
Comment (abenetopoulos) [inline text/0101-object.md:182]: 🔥

File: text/0101-object.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+The design of invariant pointers forms the basis for sharing and the global address space. Other
+invariant designs (fat pointers) have problems.
+
+The twizzler-object crate is small, and provides only a limited view of objects. Any part of an
+object that can mutate cannot be exposed by the twizzler-object crate at all, even via immutable
+reference. The only exception is atomically reading an invariant pointer, but even this requires
+interpretation via the FOT to be useful, and this requires interaction with twizzler-nando to be
+safe. The twizzler-object crate *does* expose *unsafe* functions for getting access to mutable
+object memory for the purposes of implementing twizzler-nando atop twizzler-object.
```
Comment (abenetopoulos) [inline text/0101-object.md:241]: only for 🤠

**Review by abenetopoulos — APPROVED** [review]
LGTM

**Pull request merged**